File: JukeboxExtended/src/main/java/me/tajam/jext/command/ExecutorPlayMusic.java
package me.tajam.jext.command;

import me.tajam.jext.config.ConfigDiscManager;
import me.tajam.jext.DiscContainer;
import me.tajam.jext.DiscPlayer;
import me.tajam.jext.SMS;

import java.util.List;

import org.bukkit.SoundCategory;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

class ExecutorPlayMusic extends ExecutorAdapter {

  ExecutorPlayMusic() {
    super("playmusic");
    addParameter(new Parameter("player", new CompletorPlayer()));
    addParameter(new Parameter("namespace", new CompletorDisc()));
    addParameter(new Parameter("pitch", new CompletorFloat(1.0f, 1.5f, 0.5f), false));
    addParameter(new Parameter("volume", new CompletorFloat(4.0f, 1.0f, 0.5f), false));
  }

  @Override
  boolean executePlayer(Player sender, String[] args) {
    return mergedExecute(sender, args);
  }

  @Override
  boolean executeCommand(CommandSender sender, String[] args) {
    return mergedExecute(sender, args);
  }

  private boolean mergedExecute(CommandSender sender, String[] args) {
    final List<Player> players = new PlayerSelector(sender, args[0]).getPlayers();
    if (players == null) return true;

    final DiscContainer disc = ConfigDiscManager.getInstance().getDisc(args[1]);
    if (disc == null) {
      new SMS().eror().t("Disc with the namespace ").o(args[1]).t(" doesn't exists.").send(sender);
      return true;
    }
    final DiscPlayer discPlayer = new DiscPlayer(disc);

    float pitch = 1.0f;
    if (args.length >= 3) {
      try {
        pitch = Float.parseFloat(args[2]);
      } catch (NumberFormatException e) {
        new SMS().eror().t("Wrong number format for pitch parameter.");
        return true;
      }
    }

    boolean isMusic = true;
    if (args.length >= 4) {
      isMusic = false;
      try {
        final float volume = Float.parseFloat(args[3]);
        discPlayer.setPitch(pitch);
        discPlayer.setVolume(volume);
      } catch (NumberFormatException e) {
        new SMS().eror().t("Wrong number format for volume parameter.");
        return true;
      }
    }

    for (Player player : players) {
      if (isMusic) {
        player.playSound(player.getLocation(), disc.getNamespace(), SoundCategory.RECORDS, Float.MAX_VALUE, pitch);
        new SMS().info().t("Music ").p().t(" is playing now.").send(player, disc);
      } else {
        discPlayer.play(player.getLocation());
      }
    }
    
    final Integer playerCount = players.size();
    if (playerCount >= 2) {
      new SMS().warn().t("Played music ").o().t(" to ").o().t(" players!").send(sender, disc, playerCount);
    } else if (playerCount == 1) {
      new SMS().okay().t("Played music ").o().t(" to ").o(players.get(0).getDisplayName()).t(".").send(sender, disc);
    } else {
      new SMS().eror().t("Played music to no player, something might when wrong!").send(sender);
    }
    return true;
  }
}

File: src/main/java/me/spartacus04/jext/commands/ExecutorPlayMusic.kt
package me.spartacus04.jext.commands

import me.spartacus04.jext.commands.adapter.ExecutorAdapter
import me.spartacus04.jext.commands.adapter.ParameterDisc
import me.spartacus04.jext.commands.adapter.ParameterNumber
import me.spartacus04.jext.commands.adapter.ParameterPlayer
import me.spartacus04.jext.utils.sendJEXTMessage
import org.bukkit.*
import org.bukkit.command.CommandSender

/**
 * ExecutorPlayMusic is a class used to register the "playmusic" command to the plugin.
 *
 * @constructor The constructor is empty because the class does not have any properties.
 */
internal class ExecutorPlayMusic : ExecutorAdapter("playmusic") {
    init {
        addParameter(ParameterPlayer(true))
        addParameter(ParameterDisc(true))
        addParameter(ParameterNumber(false, 0.5f, 1.0f, 1.5f).setName("pitch"))
        addParameter(ParameterNumber(false, 4.0f, 1.0f, 0.5f).setName("volume"))
    }

    /**
     * The function `execute` plays a disc to all selected players.
     *
     * @param sender The player who executed the command.
     * @param args The arguments that were passed to the command.
     */
    override fun execute(sender: CommandSender, args: Array<String>) {
        val players = ParameterPlayer.getPlayers(args[0], sender)

        val disc = ParameterDisc.getDisc(args[1])
        if (disc == null) {
            sender.sendJEXTMessage("disc-namespace-not-found", hashMapOf(
                "namespace" to args[1]
            ))

            return
        }

        var pitch = 1.0f
        if (args.size >= 3) {
            pitch = try {
                args[2].toFloat()
            } catch (e: NumberFormatException) {
                sender.sendJEXTMessage("wrong-number-format", hashMapOf(
                    "param" to "pitch"
                ))

                return
            }
        }

        var isMusic = true

        var volume = 4.0f
        if (args.size >= 4) {
            isMusic = false

            try {
                volume = args[3].toFloat()
            } catch (e: NumberFormatException) {
                sender.sendJEXTMessage("wrong-number-format", hashMapOf(
                    "param" to "volume"
                ))

                return
            }
        }

        for (player in players) {
            if (isMusic) {
                player.playSound(player.location, disc.namespace, SoundCategory.RECORDS, Float.MAX_VALUE, pitch)
                sender.sendJEXTMessage("music-now-playing", hashMapOf(
                    "name" to disc.displayName
                ))
            } else {
                disc.play(player, pitch, volume)
            }
        }

        val playerCount = players.size

        if (playerCount >= 2) {
            sender.sendJEXTMessage("played-music-to-multiple", hashMapOf(
                "name" to disc.namespace,
                "playercount" to playerCount.toString()
            ))
        } else if (playerCount == 1) {
            sender.sendJEXTMessage("played-music-to", hashMapOf(
                "name" to disc.displayName,
                "player" to players[0].name
            ))
        } else {
            sender.sendJEXTMessage("played-music-to-no-one")
        }
    }
}